### ISSUE TITLE
Missing "Link" Header for Critical Resources in Production in Next.js 15

### ISSUE THREAD

**Issue by stefa98** [vercel/next.js#71722]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/next-js-15-d55c3c

### To Reproduce

1. Run the application in development mode (next dev).
2. Check the response headers: the Link header for critical resources (e.g., fonts, styles) is present.
3. Build the application for production (next build and next start) or deploy it to Vercel in preview or production mode.
4. In production, the Link header is missing.

### Current vs. Expected behavior

- Current behavior:
In development mode, the Link header correctly includes critical resources to preload, such as fonts and stylesheets. However, in production or preview (e.g., on Vercel), the Link header is not present, and resources are not preloaded in the same way.

- Expected behavior:
The Link header should also be present in production mode, allowing the browser to preload critical resources and potentially enable optimizations like 103 Early Hints.

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Pro
  Available memory (MB): 32547
  Available CPU cores: 20
Binaries:
  Node: 20.11.1
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 15.0.1 // Latest available version is detected (15.0.1).
  eslint-config-next: 15.0.1
  react: 19.0.0-rc-69d4b800-20241021
  react-dom: 19.0.0-rc-69d4b800-20241021
  typescript: 5.6.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Output (export/standalone), Performance

### Which stage(s) are affected? (Select all that apply)

next build (local), next start (local), Vercel (Deployed)

### Additional context

**The problem also occurs with the default nextjs template, implemented with “npx create-next-app”**

I believe I found the logic that handles critical resources in the Link header in the following file:

https://github.com/vercel/next.js/blob/51d6e76596f42ea51dad6b4bf6fb7817d91a79ec/packages/next/src/compiled/react-dom-experimental/cjs/react-dom-server.node.production.js#L5723

The function safelyEmitEarlyPreloads seems to generate the Link header correctly during rendering, but it doesn't appear to be emitted in production. Is there a new option or configuration that needs to be enabled to have these critical resources included in the Link header during production? This would be particularly useful for enabling 103 Early Hints.

Thank you for any insights!

**Comment by lubieowoce** [comment]
the sandbox URL wasn't detected because it was missing `https://`, reopening

**Comment by samcx** [comment]
@stefa98 Thank you for submitting an issue!

We are taking a look <picture data-single-emoji=":one-eye-cowboy:" title=":one-eye-cowboy:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/one-eye-cowboy/38bc495b31f0fc93.png" alt=":one-eye-cowboy:" align="absmiddle"></picture>

**Comment by samcx** [comment]
@stefa98 I believe this was always the case (pre v15 stable release). Currently, you need to have [PPR](https://nextjs.org/docs/app/building-your-application/rendering/partial-prerendering) (canary only) enabled to see Link Headers in Production.

Not sure yet about the behavior in Development where it's showing without PPR enabled, so sharing this with the team!

**Comment by childrentime** [comment]
https://github.com/vercel/next.js/discussions/36089 To be honest, I eagerly hope that Next.js can support this feature. The neighboring Nuxt.js (https://nuxt.com/docs/guide/going-further/experimental-features#writeearlyhints) has already supported it for a while.

**Comment by stefa98** [comment]
Hi @samcx 

Is there any news. about it? It would be a great thing to implement “Link” Headers for Critical Resources. it is a standard now that is used by so many websites and allows for better performance. Also I don't think it is that complicated to implement, it would be very nice for the next release and besides it is an issue that has been presented for months by the community

**Comment by amitkoth** [comment]
Agreed. All that's needed is a proper header, and something like Cloudflare Pages _automatically_ emits Early Hints:
https://developers.cloudflare.com/pages/configuration/early-hints/#2-automatic-link-header-generation

**Comment by wyattjoh** [comment]
The browser can prioritize better when there are link tags present in the page versus the header. When the page is fully static, there's no time advantage to injecting the link headers. It's present in dev all the time because these pages are rendered dynamically each time, though we may be omitting that in future releases to make it more consistent.

When Partial Prerendering is enabled (via the `ppr: true` flag), we do emit the link headers because we sometimes may not be able to serve the link tags early, and as a result, there is a time advantage to sending the header.

Future work may enable the 103 early hints, but we don't have a timeline for that yet. (but stay tuned 🥳!)

TLDR; this is working as expected, so we're going to close this for now. I'll move this to a discussion so we can continue to receive feedback on this case.
